### PR TITLE
Decent implementation of Klipper_led_effects and rainbow barf lets

### PR DESCRIPTION
**Pull request by claudioguareschi**
I took the liberty to modify your excellent klipain to add support for led_effects and rainbow barf.
Led_effects can be set separately for caseligts, plain sb status leds/rainbow barf in any combination.
Respecting yout phylosophy configuration is done in printer.cfg
Few files were created in config/hardware/lights to support rainbow barf and led effects.
Logic in mcros/hardware_functions/status_leds.cfg was adapted to support led effects, attempting to respect the existing code as much as possible.
Leveling and meshing functions were modified to support led status changes when run from mainsail/fluidd

I played with the effects a little bit, tested using plain sb 3 led setup (I have a barf coming soon). The effects I choose respecting the existing default colors as much as possible are ok for now but need to be refined to make the final setup look really good. Creating the right default effects is a work in progress.

Hope this is useful. Let me know what you think

Claudio

### CODE REVIEW

**Comment by Surion79** [comment]
that looks pretty nice on first flythrough :)

**Comment by Benoitone** [comment]
I'm very interested to add this future too for the next ERCF 🤩

File: macros/hardware_functions/status_leds.cfg [hunk 1]
```diff
@@ -180,12 +180,23 @@ gcode:
...
 
     {% if printer["gcode_macro _USER_VARIABLES"].status_leds_enabled %}
-        _SET_LEDS_BY_NAME LEDS="logo" COLOR={status_color[color].logo} TRANSMIT={logo_transmit}
-        _SET_LEDS_BY_NAME LEDS="nozzle" COLOR={status_color[color].nozzle} TRANSMIT=1
+        {% if printer["gcode_macro _USER_VARIABLES"].status_leds_logo_led_name == "status_leds"%}
+            _SET_LEDS_BY_NAME LEDS="logo" COLOR={status_color[color].logo} TRANSMIT={logo_transmit}
+            _SET_LEDS_BY_NAME LEDS="nozzle" COLOR={status_color[color].nozzle} TRANSMIT=1
+        {% elif printer["gcode_macro _USER_VARIABLES"].status_leds_logo_led_name == "status_leds_effects"%}
+            {% set print_line = "Effect changed - logo: %s, nozzle: %s\n" % ("sb_logo_" + status_color[color].logo, "sb_nozzle_" + status_color[color].nozzle) %}
+            {action_respond_info(print_line)}
```
Comment (Surion79) [inline macros/hardware_functions/status_leds.cfg:188]: is this needed? looks like some sort of info printing of the print?

File: user_templates/printer.cfg [hunk 2]
```diff
@@ -154,8 +154,18 @@
...
+# uncomment the following line to enable neopixel caselights with led effects
+# [include config/hardware/lights/neopixel_caselight_effects.cfg]
+# uncomment the following line to enable stealthburner standard 3 rgb leds
 # [include config/hardware/lights/status_leds.cfg]
+# uncomment the following line to enable stealthburner standard 3 rgb leds with led effects
+# [include config/hardware/lights/status_leds_effects.cfg]
+# uncomment the following line to enable stealthburner rainbow barf 10 rgb leds
+# [include config/hardware/lights/status_leds_rb.cfg]
+# uncomment the following line to enable stealthburner rainbow barf 10 rgb leds with led effects
+# [include config/hardware/lights/status_leds_rb_effects.cfg]
```
Comment (Surion79) [inline user_templates/printer.cfg:168]: @Benoitone @Frix-x i am unsure about this includes. maybe group them for led effects and one without led effects? or even have a single include for led effects? i am not sure if users would want to have one part led effect and the other not.

File: macros/hardware_functions/status_leds.cfg (hunk 1 above)
Comment (Surion79) [reply macros/hardware_functions/status_leds.cfg:188]: (i mean line 188)

Comment (claudioguareschi) [reply macros/hardware_functions/status_leds.cfg:188]: No I wouldn't have it on the finished implementation. It is there more for debug purposes and to make sure effects are triggered when they are supposed to. For example QGL effect wasn't working for me and printing out this info allowed me to find a reset to the ready state in conditional homing (_CG28 macro) outside its if condition that was messing things up. It was making the regular status led flicker and totally messing up the effects. This implementation is not completed yet. I just submitted to see if you guys were interested in me continuing pursuing this avenue, as I saw other people already attempting to integrate led effects into klippain.

File: user_templates/printer.cfg (hunk 2 above)
Comment (claudioguareschi) [reply user_templates/printer.cfg:168]: I agree with you that would b ideal to have a less cluttered session, but I haven't found another way. Here we are jostling 3 things, actually 4, that may need to have effects attached: Caselights, minidisplay, and stealthburner 3leds  and rainbow barf variant. Unfortunately the effect definition doesn't accept variables for the leds settings i.e.:

[led_effect sb_logo_busy]
autostart:              false
frame_rate:             24
leds:
    neopixel:sb_leds (1-8).  ### I wish I could use a variable or alias here ###
layers:
    breathing  3 1 top (1,0,0)

This forced me to use all those includes to declare the effects literally and I left out effects for mini display not to clutter printer.cfg even further. In the end the user has the choice to use effects wherever he wants. There may be a use case where people have a ton (>100) leds for case lights and may want to use effects only on stealth burner. Never happened to me (I use 84 case lights leds) but somebody reported QGL hesitating /stopping with many leds and complex effects. If anybody has a suggestion on how to rework this to make it more palatable feel free...

Comment (Frix-x) [reply user_templates/printer.cfg:168]: I agree that grouping into two groups of includes would be better. Something like this for example:

```
# [include config/hardware/lights/fcob_white.cfg]
# [include config/hardware/lights/neopixel_caselight.cfg]
# [include config/hardware/lights/status_leds.cfg] # Standard StealthBurner style LEDs
# [include config/hardware/lights/status_leds_rb.cfg] # Rainbow Barf StealthBurner style LEDs

# The following files are to be used with the LED effect plugin: https://github.com/julianschill/klipper-led_effect
# [include config/hardware/lights/neopixel_caselight_effects.cfg]
# [include config/hardware/lights/status_leds_effects.cfg] # Standard StealthBurner style LEDs
# [include config/hardware/lights/status_leds_rb_effects.cfg] # Rainbow Barf StealthBurner style LEDs
```

File: macros/miscs/startup.cfg [hunk 3]
```diff
@@ -40,6 +43,10 @@ gcode:
     _INIT_USERCUSTOM
 
     RESPOND MSG="Klippain started and ready !"
+    ## Set the status LED to ready
+    {% if status_leds_enabled %}
+        STATUS_LEDS COLOR="READY"
+    {% endif %}
```
Comment (Frix-x) [inline macros/miscs/startup.cfg:49]: Very good idea! But can you move it into its own `_INIT_LEDS` macro please?

File: macros/hardware_functions/status_leds.cfg (hunk 1 above)
Comment (Frix-x) [reply macros/hardware_functions/status_leds.cfg:188]: Yes, I'm all in for this PR 🚀 
Thanks a lot for your work!

**Review by Frix-x — CHANGES_REQUESTED** [review]
Ok thank you very much for this big PR. I'm really happy about it and it's already pretty good.
I've made some comments to make it a little bit better on the user side.

Also some more comments here:
1. Please rename all "rb" references in the filenames, etc... to something more easy to understand like "rainbowbarf" or at least "rbarf"
2. You've made some change about the placement of the STATUS_LED call in the submacros. This is pretty good and a work that has already started with #404 . But I'll close #404 in favor of your PR that is more complete. In order to clean up a little bit and prepare for a future merge, can you remove the comments you've left at the previous places?

**Comment by Surion79** [comment]
i really like the quality of the PR. It is indeed decent and thought through.

**Comment by claudioguareschi** [comment]
Be glad to make all the proposed fixes. Easily done. 
I would love for somebody with a better aesthetic sense than I have to help out with some thought out effects that make the printer look good without making it look like a flying saucer!
A couple of questions:
Should we consider additional printer states for heater bed soaking and/or ercf functions?

**Comment by Surion79** [comment]
>  A couple of questions: Should we consider additional printer states for heater bed soaking and/or ercf functions?

Regarding ERCF/HH, Benetoine is currently integrating an updated version. So I think for now, we should go forward with this
Regarding additional printer states: It is important to understand, when which lights are triggered. If someone is not careful, the status might result in strobe flashes (because there is nothing done between two led status calls, thats why _CG28 got changed recently)

**Comment by Frix-x** [comment]
> Should we consider additional printer states for heater bed soaking and/or ercf functions?

I think this is not needed for now. We can keep it like this since all the previous state are correctly managed.

On the ERCF topic, there is a big ERCF PR (#371) ongoing to get compatibility with the ERCFv2 and the latest HH software. So I think the ERCF states could be managed directly in this PR when your PR is merged. Feel free to participate to it if you want :)

**Comment by Surion79** [comment]
@Frix-x the base must be changed to develop in this PR. since a lot of effort is done already, i cannot grasp possible issues on rebase to develop. currently it is pointing to main

**Comment by Benoitone** [comment]
> > Should we consider additional printer states for heater bed soaking and/or ercf functions?
> 
> I think this is not needed for now. We can keep it like this since all the previous state are correctly managed.
> 
> On the ERCF topic, there is a big ERCF PR (#371) ongoing to get compatibility with the ERCFv2 and the latest HH software. So I think the ERCF states could be managed directly in this PR when your PR is merged. Feel free to participate to it if you want :)

For the MMU/ERCF klipper_led seems to be natively supported by Happy_Hare... I haven't been able to test it yet... I am in the process of assembling my ERCFv2...
But this PR interests me a lot for the management of the other LEDs and to see how it behaves.
I would be in favor of pushing it on a new branch of klippain before its merge when @Frix-x  wishes. So that I can see what it could give in parallel with the management of Happy_Hare.

**Comment by Benoitone** [comment]
As an off topic, I still have a problem with LED flashing at the time of the QGL in _MODULE_TILTING I think there is something to check on this side...

**Comment by Surion79** [comment]
> > > Should we consider additional printer states for heater bed soaking and/or ercf functions?
> > 
> > 
> > I think this is not needed for now. We can keep it like this since all the previous state are correctly managed.
> > On the ERCF topic, there is a big ERCF PR (#371) ongoing to get compatibility with the ERCFv2 and the latest HH software. So I think the ERCF states could be managed directly in this PR when your PR is merged. Feel free to participate to it if you want :)
> 
> For the MMU/ERCF klipper_led seems to be natively supported by Happy_Hare... I haven't been able to test it yet... I am in the process of assembling my ERCFv2... But this PR interests me a lot for the management of the other LEDs and to see how it behaves. I would be in favor of pushing it on a new branch of klippain before its merge when @Frix-x wishes. So that I can see what it could give in parallel with the management of Happy_Hare.

i suggest to complete the big PR first before sidetracking it for another feature, IMHO. Ex manager had a saying: better metal on the road than gold in the storage (meaning software)

**Comment by Surion79** [comment]
> As an off topic, I still have a problem with LED flashing at the time of the QGL in _MODULE_TILTING I think there is something to check on this side...

at some point, it might need to be required to use params to differ between mainsail call and call within a process (like start_print) to correctly set the leds in manual use and automated/integrated use. I had the thought already couple time when looking at the strobe flashing issue.

**Comment by Frix-x** [comment]
Ok so I indeed changed the target to develop and will do a final review soon but it look like ok for me on a first sight now :)

**Comment by Frix-x** [comment]
Mhmm I think the automated target change directly on Github was not a good idea after this number of commit and I should have done it manually... I'll fix it.

**Pull request merged**